ENH: Add a `mamba` plugin

HaoZeke opened the pull request:
This is much faster than the `conda` solver and is more robust with a better Pythonic API. Closes #962.

HaoZeke commented:
Expects #1237

zacharyburnett commented:
@HaoZeke nice PR, this should speed up the benchmarking process! You can also try using the https://github.com/marketplace/actions/provision-with-micromamba action in the CI to make it even faster

HaoZeke commented:
> @HaoZeke nice PR, this should speed up the benchmarking process! You can also try using the https://github.com/marketplace/actions/provision-with-micromamba action to make the CI even faster

Yup, however, I think I'd need to add another plugin for `micromamba` since it isn't covered by the `python` API used here.

HaoZeke commented:
Installing dependencies on `pypy-3.7` leads to an obscure error:

```
Installing collected packages: sphinx_bootstrap_theme, sortedcontainers, snowballstemmer, sgmllib3k, pytz, python-hglib, mccabe, distlib, charset-normalizer, zipp, urllib3, typing-extensions, tomli, sphinxcontrib-serializinghtml, sphinxcontrib-qthelp, sphinxcontrib-jsmath, sphinxcontrib-htmlhelp, sphinxcontrib-devhelp, sphinxcontrib-applehelp, sniffio, pyyaml, PySocks, Pygments, pyflakes, pycodestyle, packaging, numpy, MarkupSafe, isort, iniconfig, imagesize, idna, filelock, feedparser, execnet, exceptiongroup, docutils, certifi, babel, attrs, async-generator, alabaster, requests, platformdirs, outcome, Jinja2, importlib-metadata, h11, wsproto, virtualenv, trio, sphinx, pluggy, flake8, trio-websocket, pytest, selenium, pytest-xdist, pytest-timeout, pytest-rerunfailures
ERROR: Could not install packages due to an OSError: [Errno 2] No such file or directory: '/usr/share/miniconda/envs/test/site-packages/../bin/rst2html.py'

Error: Process completed with exit code 1.
```

Dropping support for `3.7` in general might be a good idea? Especially since newer plugins (e.g. `mamba`, `memray`) require newer versions as well.

mattip commented on asv/environment.py:
```diff
@@ -428,18 +438,24 @@ def get_environment_class(conf, python):
     raise EnvironmentUnavailable(
-        "No way to create environment for python='{0}'".format(python))
+        "No way to create environment for python='{0}'".format(python)
+    )
```
f-strings might be nicer

mattip commented on asv/environment.py:
```diff
@@ -428,18 +438,24 @@ def get_environment_class(conf, python):
+        if "mamba" in environment_type and (
+            sys.version_info.major >= 3 and sys.version_info.minor < 8
+        ):
+            environment_type = "conda"
```
why the upper limit? Also: what happens if "conda" is used instead of "mamba"?

HaoZeke replied:
If `conda` is used it continues to use the old solver. The idea here was that it should silently fall back to the `conda` mechanism before `libmambapy` is present. The upper limit shouldn't be there, sorry.

mattip reviewed (commented):
It is much easier to review if the formatting changes are done in a separate PR, and not mixed in with code changes.

Rather than hard-code `pip` to install the project, shouldn't there be a way for the project to say what backend to use?

fangchenli commented on .github/workflows/ci.yml:
```diff
@@ -23,20 +23,26 @@ jobs:
 
-      - name: Set up Python ${{ matrix.python-version }}
-        uses: actions/setup-python@v4
+      - name: Setup Conda with ${{ matrix.python-version }}
+        uses: conda-incubator/setup-miniconda@v2
```
Setting up conda and memba in the CI takes longer than using `setup-python`. For a library with very few dependencies, it's not worth it. Also, the CI changes are unrelated to the topic of this PR and should be addressed separately.

HaoZeke replied:
Makes sense, originally I thought of testing the new environment handler in CI as well, which is why I wanted to add it here, but I guess that could be elsewhere.

HaoZeke commented:
Some benchmarks for cold starts `time asv run` with `.asv` removed:

```bash
# Conda
asv run  63.61s user 4.06s system 101% cpu 1:06.50 total
# Mamba
asv run  13.42s user 2.43s system 120% cpu 13.152 total
# Virtualenv
asv run  8.35s user 1.83s system 116% cpu 8.751 total
```

For this [toy project](https://github.com:HaoZeke/asv_experiments)

HaoZeke commented:
> It is much easier to review if the formatting changes are done in a separate PR, and not mixed in with code changes.
Sorry, reverted those.

> Rather than hard-code `pip` to install the project, shouldn't there be a way for the project to say what backend to use?

The project is installed with the `install_command` configuration option, the plugin only uses `pip` for the dependencies..

HaoZeke commented:
I have cleaned this up and simplified it. There are no tests, other than manual ones. Tests would require changes to the CI which can be done elsewhere.

zacharyburnett commented on asv/environment.py:
```diff
@@ -428,7 +428,7 @@ def get_environment_class(conf, python):
             return cls
     raise EnvironmentUnavailable(
-        "No way to create environment for python='{0}'".format(python))
+        "No way to create environment for python='{python}'")
```
does this need an `f`? You could use `flynt` or something to do these automatically (maybe out of scope for this PR)

HaoZeke replied:
Yup, good eye. Also I'll open another PR for `flynt` but I think @jarrodmillman has converted a lot of it already.

mattip commented:
Thanks @HaoZeke
